File: roms/gaplus.py
#
#	Gaplus
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    prg1 = z.read('gp2-4.8d') + z.read('gp2-3b.8c') + z.read('gp2-2b.8b')
    prg2 = z.read('gp2-8.11d') + z.read('gp2-7.11c') + z.read('gp2-6.11b')
    prg3 = z.read('gp2-1.4b')
    bg = z.read('gp2-5.8s')
    obj = z.read('gp2-11.11p') + z.read('gp2-10.11n') + z.read('gp2-9.11m') + z.read('gp2-12.11r')
    red = z.read('gp2-3.1p')
    green = z.read('gp2-1.1n')
    blue = z.read('gp2-2.2n')
    bgcolor = z.read('gp2-7.6s')
    objcolor_l = z.read('gp2-6.6p')
    objcolor_h = z.read('gp2-5.6n')
    snd = z.read('gp2-4.3f')
with ZipFile(argv[2]) as z:
    prg = z.read('62xx.bin')

rom = prg1 + prg2 + prg3 + bg + obj + red + green + blue + bgcolor + objcolor_l + objcolor_h + snd + prg

def pngstring(a):
    w = 1024
    img = Image.new('L', (w, ceil(len(a) / w)))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[3], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/gaplus.js
/*
 *	Gaplus
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};

let z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg1 = cat(z.read('gp2-4.8d'), z.read('gp2-3b.8c'), z.read('gp2-2b.8b'));
const prg2 = cat(z.read('gp2-8.11d'), z.read('gp2-7.11c'), z.read('gp2-6.11b'));
const prg3 = z.read('gp2-1.4b');
const bg = z.read('gp2-5.8s');
const obj = cat(z.read('gp2-11.11p'), z.read('gp2-10.11n'), z.read('gp2-9.11m'), z.read('gp2-12.11r'));
const red = z.read('gp2-3.1p');
const green = z.read('gp2-1.1n');
const blue = z.read('gp2-2.2n');
const bgcolor = z.read('gp2-7.6s');
const objcolor_l = z.read('gp2-6.6p');
const objcolor_h = z.read('gp2-5.6n');
const snd = z.read('gp2-4.3f');
z = new Zlib.Unzip(fs.readFileSync(process.argv[3]));
const prg = z.read('62xx.bin');

const rom = cat(prg1, prg2, prg3, bg, obj, red, green, blue, bgcolor, objcolor_l, objcolor_h, snd, prg);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[4], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
